crypto/src/lib.rs

```diff
@@ -0,0 +1,117 @@
+/*++
+Licensed under the Apache-2.0 license.
+Abstract:
+    Generic trait definition of Cryptographic functions.
+--*/
+
+#[cfg(feature = "openssl")]
+pub use crate::openssl::*;
+
+#[cfg(feature = "openssl")]
+pub mod openssl;
+
+#[derive(Debug, Clone, Copy)]
+pub enum AlgLen {
+    Bit256,
+    Bit384,
+}
+
+impl AlgLen {
+    pub const fn size(self) -> usize {
+        match self {
+            AlgLen::Bit256 => 256 / 8,
+            AlgLen::Bit384 => 384 / 8,
+        }
+    }
+}
+
+pub trait Hasher: Sized {
+    /// Adds a chunk to the running hash.
+    ///
+    /// # Arguments
+    ///
+    /// * `bytes` - Value to add to hash.
+    fn update(&mut self, bytes: &[u8]) -> Option<()>;
+
+    /// Finish a running hash operation and return the result.
+    ///
+    /// Once this function has been called, the object can no longer be used and
+    /// a new one must be created to hash more data.
+    ///
+    /// # Arguments
+    ///
+    /// * `digest` - Where the computed digest should be written.
+    fn finish<const SIZE: usize>(self) -> Option<[u8; SIZE]>;
+}
+
+pub trait Crypto {
+    type Cdi;
+    type Hasher: Hasher;
+
+    /// Fills the buffer with random values.
+    ///
+    /// # Arguments
+    ///
+    /// * `dst` - The buffer to be filled.
+    fn rand_bytes(dst: &mut [u8]) -> Option<()>;
+
+    /// Cryptographically hashes the given buffer.
+    ///
+    /// # Arguments
+    ///
+    /// * `profile` - Which profile is being used. This will tell the platform which algorithm to
+    ///   use.
+    /// * `bytes` - Value to be hashed.
+    /// * `digest` - Where the computed digest should be written.
+    fn hash<const SIZE: usize>(algs: AlgLen, bytes: &[u8]) -> Option<[u8; SIZE]> {
```

Comment:
I'm not sure if I have a good suggestion, but it feels a big weird to pass both SIZE and an AlgLen since they're expected to always be the same.

Reply:
That's a good point. I think I might be able to figure out a way to get rid of that enum and just use the generic. WDYT?

This'll probably have to wait till Monday though.

Reply:
I think just using the generic makes sense. No rush!

Reply:
Enum const generics aren't supported by stable quite yet. I think it might actually be best to go back to using an enum. I think that is more obvious about how to use it and it is the same convention as OpenSSL.